File: sol/2182/2182.cpp
struct Solution {
    string repeatLimitedString(const string &s, int repeatLimit) {
        int freq[26] = {0};
        for (char c : s){
            ++freq[c - 'a'];
        }
        vector<pair<char, int>> st;
        for (int i = 0; i < 26; ++i){
            if (freq[i] != 0){
                st.push_back({'a' + i, freq[i]});
            }
        }
        string res = "";
        while (!st.empty()){
            auto [c, f] = st.back();
            st.pop_back();
            int take = min(f, repeatLimit);
            res.append(take, c);
            f -= take;
            if (f != 0 && !st.empty()){
                auto [cc, ff] = st.back();
                st.pop_back();
                res += cc;
                if (ff != 1){
                    st.push_back({cc, ff - 1});
                }
                st.push_back({c, f});
            }
        }
        return res;
    }
};

File: sol/2182/2182.py
class Solution:
    def repeatLimitedString(self, s: str, repeatLimit: int) -> str:
        freq = [0] * 26
        for c in s:
            freq[ord(c) - ord('a')] += 1
        st = sorted(((chr(i + ord('a')), f) for i, f in enumerate(freq) if f != 0))
        res = ""
        while len(st) != 0:
            c, f = st.pop()
            take = min(f, repeatLimit)
            res += c * take
            f -= take
            if f != 0 and len(st) != 0:
                cc, ff = st.pop()
                res += cc
                if ff != 1:
                    st.append((cc, ff - 1))
                st.append((c, f))
        return res
